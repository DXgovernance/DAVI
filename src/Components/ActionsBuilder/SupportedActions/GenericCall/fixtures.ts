import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'Components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const genericDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.GENERIC_CALL,
  function: ERC20Contract.getFunction('increaseAllowance'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    param1: BigNumber.from(10000),
    param2: BigNumber.from(10000),
    param3: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
