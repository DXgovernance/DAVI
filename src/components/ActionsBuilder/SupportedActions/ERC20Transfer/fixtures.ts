import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const erc20TransferEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const erc20TransferDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    _to: '0xaF8eB8C3A5d9d900AA0B98e3Df0bcC17d3C5F698',
    _value: BigNumber.from('5000000000000000000'),
  },
};
