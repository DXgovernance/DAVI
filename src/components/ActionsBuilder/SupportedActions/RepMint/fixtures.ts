import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import ERC20SnapshotRep from 'contracts/ERC20SnapshotRep.json';

const ERC20SnapshotRepContract = new utils.Interface(ERC20SnapshotRep.abi);

export const repMintEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '',
  value: BigNumber.from(0),
  args: {
    to: '',
    amount: BigNumber.from(0),
  },
};

export const repMintDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
    amount: BigNumber.from(10000),
  },
};
