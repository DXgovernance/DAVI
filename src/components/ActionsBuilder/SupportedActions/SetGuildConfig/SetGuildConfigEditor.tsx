import { FC, useMemo } from 'react';
import { ActionEditorProps } from '..';
import { useGuildConfig } from 'Modules/Guilds/Hooks/useGuildConfig';
import { useTypedParams } from 'Modules/Guilds/Hooks/useTypedParams';
import { NumericalInput } from 'components/primitives/Forms/NumericalInput';
import { BigNumber } from 'ethers';
import { Button } from 'components/primitives/Button';
import { DurationInput } from 'components/primitives/Forms/DurationInput';

import {
  Control,
  ControlLabel,
  ControlRow,
} from 'components/primitives/Forms/Control';
import { useTranslation } from 'react-i18next';
import { ErrorLabel as Error } from 'components/primitives/Forms/ErrorLabel';
import { Controller, useForm } from 'react-hook-form';

const bn = (n: string | number | BigNumber) => BigNumber.from(n);

export const pickValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (!!current && BigNumber.isBigNumber(current)) {
    if (
      !!modifyed &&
      BigNumber.isBigNumber(modifyed) &&
      !bn(modifyed).eq(current)
    ) {
      return modifyed;
    }
    return current;
  } else if (!!modifyed && BigNumber.isBigNumber(modifyed)) {
    return modifyed;
  }
  return BigNumber.from(0);
};

export interface Data {
  proposalTime: BigNumber;
  timeForExecution: BigNumber;
  votingPowerForProposalExecution: BigNumber;
  votingPowerForProposalCreation: BigNumber;
  voteGas: BigNumber;
  maxGasPrice: BigNumber;
  maxActiveProposals: BigNumber;
  lockTime: BigNumber;
  minimumMembersForProposalCreation: BigNumber;
  minimumTokensLockedForProposalCreation: BigNumber;
}
type ControlField =
  | 'votingPowerForProposalExecution'
  | 'votingPowerForProposalCreation'
  | 'voteGas'
  | 'maxGasPrice'
  | 'maxActiveProposals'
  | 'minimumMembersForProposalCreation'
  | 'minimumTokensLockedForProposalCreation'
  | 'proposalTime'
  | 'timeForExecution'
  | 'lockTime';

export const fields = [
  { name: 'votingPowerForProposalExecution', type: 'number' },
  { name: 'votingPowerForProposalCreation', type: 'number' },
  { name: 'voteGas', type: 'number' },
  { name: 'maxGasPrice', type: 'number' },
  { name: 'maxActiveProposals', type: 'number' },
  { name: 'minimumMembersForProposalCreation', type: 'number' },
  { name: 'minimumTokensLockedForProposalCreation', type: 'number' },
  { name: 'proposalTime', type: 'duration' },
  { name: 'timeForExecution', type: 'duration' },
  { name: 'lockTime', type: 'duration' },
];

const SetGuildConfigEditor: FC<ActionEditorProps> = ({
  decodedCall,
  onSubmit,
}) => {
  const { guildId } = useTypedParams();
  const { data: currentGuildConfig } = useGuildConfig(guildId);
  const { t } = useTranslation();

  const parsedData = useMemo<Data>(() => {
    if (!decodedCall) return null;
    const {
      _proposalTime: decodedCallProposalTime,
      _timeForExecution: decodedCallTimeForExecution,
      _votingPowerForProposalExecution:
        decodedCallVotingPowerForProposalExecution,
      _votingPowerForProposalCreation:
        decodedCallVotingPowerForProposalCreation,
      _voteGas: decodedCallVoteGas,
      _maxGasPrice: decodedCallMaxGasPrice,
      _maxActiveProposals: decodedCallMaxActiveProposals,
      _lockTime: decodedCallLockTime,
      _minimumMembersForProposalCreation:
        decodedCallMinimumMembersForProposalCreation,
      _minimumTokensLockedForProposalCreation:
        decodedCallMinimumTokensLockedForProposalCreation,
    } = decodedCall.args;

    const {
      proposalTime: currentproposalTime,
      timeForExecution: currenttimeForExecution,
      votingPowerForProposalExecution: currentvotingPowerForProposalExecution,
      votingPowerForProposalCreation: currentvotingPowerForProposalCreation,
      voteGas: currentVoteGas,
      maxGasPrice: currentMaxGasPrice,
      maxActiveProposals: currentmaxActiveProposals,
      lockTime: currentlockTime,
      minimumMembersForProposalCreation:
        currentMinimumMembersForProposalCreation,
      minimumTokensLockedForProposalCreation:
        currentMinimumTokensLockedForProposalCreation,
    } = currentGuildConfig;

    // const { asset, functionName, tab } = decodedCall.optionalProps;
    return {
      proposalTime: pickValue(currentproposalTime, decodedCallProposalTime),
      timeForExecution: pickValue(
        currenttimeForExecution,
        decodedCallTimeForExecution
      ),
      votingPowerForProposalExecution: pickValue(
        currentvotingPowerForProposalExecution,
        decodedCallVotingPowerForProposalExecution
      ),
      votingPowerForProposalCreation: pickValue(
        currentvotingPowerForProposalCreation,
        decodedCallVotingPowerForProposalCreation
      ),
      voteGas: pickValue(currentVoteGas, decodedCallVoteGas),
      maxGasPrice: pickValue(currentMaxGasPrice, decodedCallMaxGasPrice),
      maxActiveProposals: pickValue(
        currentmaxActiveProposals,
        decodedCallMaxActiveProposals
      ),
      lockTime: pickValue(currentlockTime, decodedCallLockTime),
      minimumMembersForProposalCreation: pickValue(
        currentMinimumMembersForProposalCreation,
        decodedCallMinimumMembersForProposalCreation
      ),
      minimumTokensLockedForProposalCreation: pickValue(
        currentMinimumTokensLockedForProposalCreation,
        decodedCallMinimumTokensLockedForProposalCreation
      ),
    };
  }, [decodedCall, currentGuildConfig]);

  const { control, handleSubmit } = useForm({
    // resolver: validateSetPermissions,
    // context: { t, activeTab },
    defaultValues: {
      proposalTime: parsedData.proposalTime,
      timeForExecution: parsedData.timeForExecution,
      votingPowerForProposalExecution:
        parsedData.votingPowerForProposalExecution,
      votingPowerForProposalCreation: parsedData.votingPowerForProposalCreation,
      voteGas: parsedData.voteGas,
      maxGasPrice: parsedData.maxGasPrice,
      maxActiveProposals: parsedData.maxActiveProposals,
      lockTime: parsedData.lockTime,
      minimumMembersForProposalCreation:
        parsedData.minimumMembersForProposalCreation,
      minimumTokensLockedForProposalCreation:
        parsedData.minimumTokensLockedForProposalCreation,
    },
  });

  const submitAction = (values: Data) => {
    // maxActiveProposals,
    const {
      proposalTime,
      timeForExecution,
      votingPowerForProposalExecution,
      votingPowerForProposalCreation,
      voteGas,
      maxGasPrice,
      maxActiveProposals,
      lockTime,
      minimumMembersForProposalCreation,
      minimumTokensLockedForProposalCreation,
    } = values;
    const call = {
      ...decodedCall,
      args: {
        _proposalTime: proposalTime,
        _timeForExecution: timeForExecution,
        _votingPowerForProposalExecution: votingPowerForProposalExecution,
        _votingPowerForProposalCreation: votingPowerForProposalCreation,
        _voteGas: voteGas,
        _maxGasPrice: maxGasPrice,
        _maxActiveProposals: maxActiveProposals,
        _lockTime: lockTime,
        _minimumMembersForProposalCreation: minimumMembersForProposalCreation,
        _minimumTokensLockedForProposalCreation:
          minimumTokensLockedForProposalCreation,
      },
      // optionalProps: {
      //   updatedFields: getUpdatedValues(currentGuildConfig, values),
      // },
    };
    onSubmit(call);
  };

  return (
    <form onSubmit={handleSubmit(submitAction, console.error)}>
      {fields.map(f => {
        return (
          <div>
            <Controller
              name={f.name as ControlField}
              control={control}
              render={({ field: { ref, ...field }, fieldState }) => {
                const { invalid, error } = fieldState;
                const onChange = (value: string | number) => {
                  field.onChange(BigNumber.from(value ?? 0));
                };
                return (
                  <>
                    <Control>
                      <ControlLabel>
                        {/* {t('proposalTime', { defaultValue: field.name })}
                         */}
                        {field.name}
                      </ControlLabel>
                      <ControlRow>
                        {f.type === 'duration' ? (
                          <DurationInput
                            {...field}
                            aria-label={field.name}
                            value={Number(
                              BigNumber.from(field.value ?? 0).toString()
                            )}
                            onChange={onChange}
                          />
                        ) : (
                          <NumericalInput
                            {...field}
                            value={BigNumber.from(field.value ?? 0).toString()}
                            aria-label={field.name}
                            isInvalid={invalid && !!error}
                            onChange={onChange}
                          />
                        )}
                      </ControlRow>
                    </Control>
                    {invalid && !!error && <Error>{error.message}</Error>}
                  </>
                );
              }}
            />
          </div>
        );
      })}

      <Button m="1rem 0 0" fullWidth type="submit">
        {t('saveAction')}
      </Button>
    </form>
  );
};

export default SetGuildConfigEditor;
