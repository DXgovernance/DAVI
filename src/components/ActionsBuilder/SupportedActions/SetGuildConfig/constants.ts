export const FIELDS = [
  { name: 'proposalTime', label: 'Proposal time', type: 'duration' },
  { name: 'timeForExecution', label: 'Time for execution', type: 'duration' },
  {
    name: 'votingPowerPercentageForProposalExecution',
    label: 'Voting power percentage for proposal execution',
    type: 'percentage',
  },
  {
    name: 'votingPowerPercentageForProposalCreation',
    label: 'Voting power percentage for proposal creation',
    type: 'percentage',
  },
  { name: 'voteGas', label: 'Vote gas', type: 'number' },
  { name: 'maxGasPrice', label: 'Max gas price', type: 'number' },
  { name: 'maxActiveProposals', label: 'Max active proposals', type: 'number' },
  { name: 'lockTime', label: 'Lock time', type: 'duration' },
  {
    name: 'minimumMembersForProposalCreation',
    label: 'Minimum members for proposal creation',
    type: 'number',
  },
  {
    name: 'minimumTokensLockedForProposalCreation',
    label: 'Minimum tokens locked for proposal creation',
    type: 'number',
  },
];
