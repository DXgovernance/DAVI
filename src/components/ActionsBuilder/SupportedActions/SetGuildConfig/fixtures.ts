import { BigNumber, utils } from 'ethers';
import { SupportedAction } from 'components/ActionsBuilder/types';
import BaseERC20Guild from 'contracts/BaseERC20Guild.json';

export const mockBigNumber = BigNumber.from(100000000);
export const BaseERC20GuildContract = new utils.Interface(BaseERC20Guild.abi);

export const emptyDecodedCall = {
  from: '',
  callType: SupportedAction.SET_GUILD_CONFIG,
  function: BaseERC20GuildContract.getFunction('setConfig'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _proposalTime: '',
    _timeForExecution: '',
    _votingPowerPercentageForProposalExecution: '',
    _votingPowerPercentageForProposalCreation: '',
    _voteGas: '',
    _maxGasPrice: '',
    _maxActiveProposals: '',
    _lockTime: '',
    _minimumMembersForProposalCreation: '',
    _minimumTokensLockedForProposalCreation: '',
  },
  optionalProps: {},
};
