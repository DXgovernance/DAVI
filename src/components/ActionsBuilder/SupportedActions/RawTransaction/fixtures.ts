import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';

export const rawDataCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.RAW_TRANSACTION,
  function: null,
  to: MOCK_ADDRESS,
  value: BigNumber.from('1000000000000000000'),
  args: null,
  optionalProps: {
    data: '0x01',
  },
};
