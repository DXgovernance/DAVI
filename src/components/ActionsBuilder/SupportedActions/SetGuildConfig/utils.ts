import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
