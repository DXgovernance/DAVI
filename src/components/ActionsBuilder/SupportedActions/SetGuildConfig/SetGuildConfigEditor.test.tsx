import { render } from 'utils/tests';
import { BigNumber } from 'ethers';
import SetGuildConfigEditor from './SetGuildConfigEditor';
import { emptyDecodedCall } from './fixtures';
import { FIELDS } from './constants';

const mockBigNumber = BigNumber.from(100000000);

jest.mock('Modules/Guilds/Hooks/useGuildConfig', () => ({
  useGuildConfig: () => ({
    data: {
      permissionRegistry: '0x23120390909dq0w9adsj832921293989',
      name: 'TEST GUILD',
      proposalTime: mockBigNumber,
      timeForExecution: mockBigNumber,
      maxActiveProposals: mockBigNumber,
      votingPowerForProposalCreation: mockBigNumber,
      tokenVault: '0X9628SD7128DAS78SHDF8HF832H283FH283RH98ASHD',
      lockTime: mockBigNumber,
      voteGas: mockBigNumber,
      maxGasPrice: mockBigNumber,
      votingPowerPercentageForProposalExecution: mockBigNumber,
      votingPowerPercentageForProposalCreation: mockBigNumber,
      minimumMembersForProposalCreation: mockBigNumber,
      minimumTokensLockedForProposalCreation: mockBigNumber,
      votingPowerForProposalExecution: mockBigNumber,
      token: '0x30ii2jnn32ru8238r293ru2389ru2938ru239r23r23',
    },
  }),
}));

jest.mock('Modules/Guilds/Hooks/useTypedParams', () => ({
  useTypedParams: () => ({
    guildId: '0xE9bDaB08f2FBb370d2a6F6661a92d9B6157E9fd2',
  }),
}));

describe('SetGuildConfigEditor', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <SetGuildConfigEditor
        decodedCall={emptyDecodedCall}
        onSubmit={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('Should render all fields', () => {
    const { getByText } = render(
      <SetGuildConfigEditor
        decodedCall={emptyDecodedCall}
        onSubmit={jest.fn()}
      />
    );
    FIELDS.forEach(field => {
      expect(getByText(field.label)).toBeInTheDocument();
    });
  });
});
