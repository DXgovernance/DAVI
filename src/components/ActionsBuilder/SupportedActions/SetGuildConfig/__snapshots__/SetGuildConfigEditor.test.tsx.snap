// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetGuildConfigEditor Should match snapshot 1`] = `
.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0.75rem 0;
  width: 100%;
}

.c2 {
  margin-bottom: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  color: #A1A6B0;
  font-size: 14px;
  font-weight: 500;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c8 {
  border: none;
  background: none;
  background-color: transparent;
  box-shadow: none;
  width: 100%;
  box-sizing: border-box;
  outline: none;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  border: 1px solid #303338;
  border-radius: 1.5rem;
  padding: 0.5rem 0.8rem;
  background-color: transparent;
}

.c6 input {
  color: #fff;
  font-family: 'Inter',-apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: 500;
  text-align: left;
}

.c6 input::-webkit-input-placeholder {
  color: #A1A6B0;
}

.c6 input::-moz-placeholder {
  color: #A1A6B0;
}

.c6 input:-ms-input-placeholder {
  color: #A1A6B0;
}

.c6 input::placeholder {
  color: #A1A6B0;
}

.c6:hover {
  outline: 1px solid #fff;
}

.c6:focus {
  outline: 2px solid #fff;
}

.c7 {
  margin-right: 0.3rem;
}

.c9 {
  margin-left: 0.3rem;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  font-size: 14px;
  border: 1px solid #303338;
  background-color: transparent;
  color: #fff;
  border-radius: 32px;
  padding: 0.5rem 0.8rem;
  margin: 0.2rem;
  width: auto;
}

.c4:disabled {
  color: initial;
  opacity: 0.4;
  cursor: auto;
}

.c4:hover:enabled {
  border-color: #fff;
}

.c4:active:enabled {
  border: 1px solid #303338;
}

.c4:disabled {
  color: #303338;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  font-size: 14px;
  border: 1px solid #303338;
  background-color: #295FF4;
  color: #fff;
  border-radius: 32px;
  padding: 0.5rem 0.8rem;
  margin: 0.2rem;
  width: 100%;
  margin: 1rem 0 0;
}

.c10:disabled {
  color: initial;
  opacity: 0.4;
  cursor: auto;
}

.c10:hover:enabled {
  border-color: #fff;
}

.c10:active:enabled {
  border: 1px solid #303338;
}

.c5 {
  width: 100%;
  min-height: 37px;
  background-color: transparent;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-align: left;
  border: 1px solid inherit;
}

<div>
   
  <form>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Proposal time
        </div>
        <div
          class="c0 c3"
        >
          <button
            aria-label="Duration picker button"
            class="c4 c5"
            type="button"
          >
            3 years, 2 months, 17 days, 9 hours, 46 minutes and 40 seconds
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Time for execution
        </div>
        <div
          class="c0 c3"
        >
          <button
            aria-label="Duration picker button"
            class="c4 c5"
            type="button"
          >
            3 years, 2 months, 17 days, 9 hours, 46 minutes and 40 seconds
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Voting power percentage for proposal execution
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            >
              %
            </div>
            <input
              aria-label="Voting power percentage for proposal execution"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="votingPowerPercentageForProposalExecution"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Voting power percentage for proposal creation
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            >
              %
            </div>
            <input
              aria-label="Voting power percentage for proposal creation"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="votingPowerPercentageForProposalCreation"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Vote gas
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            />
            <input
              aria-label="Vote gas"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="voteGas"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Max gas price
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            />
            <input
              aria-label="Max gas price"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="maxGasPrice"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Max active proposals
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            />
            <input
              aria-label="Max active proposals"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="maxActiveProposals"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Lock time
        </div>
        <div
          class="c0 c3"
        >
          <button
            aria-label="Duration picker button"
            class="c4 c5"
            type="button"
          >
            3 years, 2 months, 17 days, 9 hours, 46 minutes and 40 seconds
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Minimum members for proposal creation
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            />
            <input
              aria-label="Minimum members for proposal creation"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="minimumMembersForProposalCreation"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          Minimum tokens locked for proposal creation
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
            />
            <input
              aria-label="Minimum tokens locked for proposal creation"
              autocomplete="off"
              autocorrect="off"
              class="c8"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="minimumTokensLockedForProposalCreation"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="100000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      class="c10"
      type="submit"
    >
      saveAction
    </button>
  </form>
   
</div>
`;
